lib/Core/Executor.cpp

```diff
@@ -4001,21 +4005,29 @@ void Executor::executePersistentMemoryFlush(ExecutionState &state,
       ObjectState *wos = state.addressSpace.getWriteable(mo, os);
       PersistentState *ps = dyn_cast<PersistentState>(wos);
       ref<Expr> offset = mo->getOffsetExpr(address);
-      // TODO: fork here, instead of relying on bool return value
-      // fork on "isOffsetPersisted"
-      ref<Expr> alreadyPersisted = ps->isOffsetPersisted(offset);
+
+      ref<Expr> alreadyPersisted = ps->isOffsetAlreadyPersisted(offset);
       StatePair notPersisted = fork(state, Expr::createIsZero(alreadyPersisted) , true);
       if (notPersisted.first) {
-        // offset not yet persisted, should persist it!
-        auto& goodState = *notPersisted.second;
+        auto& goodState = *notPersisted.first;
+        wos = goodState.addressSpace.getWriteable(mo, os);
+        assert(wos && "Could not get writeable ObjectState from notPersisted.first");
+        ps = dyn_cast<PersistentState>(wos);
+        assert(ps && "Could not get PersistentState from notPersisted.first");
         ps->persistCacheLineAtOffset(offset);
+        llvm::errs() << "Good Flush\n";
       }
       if (notPersisted.second) {
         // offset is already persisted, should error!
-        llvm::errs() << "Unnecessary Flush\n";
         auto& errState = *notPersisted.second;
+        wos = errState.addressSpace.getWriteable(mo, os);
```

Comment:
You should re-acqure the ObjectState from the address space, by means of 
```
const ObjectState *os = errState.addressSpace.findObject(mo);
wos = errState.addressSpace.getWriteable(mo, os);
```
at least in my mind. I could be wrong, but this seems weird otherwise.